File: docs-source/docs/modules/microservices-tutorial/examples/05-shopping-cart-service-java/src/test/java/shopping/cart/CreateTableTestUtils.java
package shopping.cart;

import static java.util.concurrent.TimeUnit.SECONDS;

import akka.actor.typed.ActorSystem;
import akka.persistence.jdbc.testkit.javadsl.SchemaUtils;
import akka.projection.jdbc.javadsl.JdbcProjection;
import java.util.Objects;
import javax.persistence.EntityManager;
import org.slf4j.LoggerFactory;
import org.springframework.orm.jpa.EntityManagerFactoryUtils;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.transaction.TransactionStatus;
import shopping.cart.repository.HibernateJdbcSession;

public class CreateTableTestUtils {

  /** 
    * Test utility to create journal and projection tables for tests environment. 
    * JPA/Hibernate tables are auto created (drop-and-create) using settings flag, see persistence-test.conf
    */
  public static void createTables(JpaTransactionManager transactionManager, ActorSystem<?> system)
      throws Exception {

    // create schemas
    // ok to block here, main test thread
    SchemaUtils.dropIfExists(system).toCompletableFuture().get(30, SECONDS);
    ;
    SchemaUtils.createIfNotExists(system).toCompletableFuture().get(30, SECONDS);

    JdbcProjection.dropOffsetTableIfExists(
            () -> new HibernateJdbcSession(transactionManager), system)
        .toCompletableFuture()
        .get(30, SECONDS);
    ;
    JdbcProjection.createOffsetTableIfNotExists(
            () -> new HibernateJdbcSession(transactionManager), system)
        .toCompletableFuture()
        .get(30, SECONDS);
    ;

    LoggerFactory.getLogger(CreateTableTestUtils.class).info("Tables created");
  }
}

File: docs-source/docs/modules/microservices-tutorial/examples/02-shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala
package shopping.cart

import akka.Done
import akka.actor.typed.ActorSystem
import akka.persistence.jdbc.testkit.scaladsl.SchemaUtils
import akka.projection.jdbc.scaladsl.JdbcProjection
import com.typesafe.config.Config
import org.slf4j.LoggerFactory
import shopping.cart.repository.{ DBsFromConfig, ScalikeJdbcSession }

import java.nio.file.Paths
import scala.concurrent.duration._
import scala.concurrent.{ Await, ExecutionContext, Future }

object CreateTableTestUtils {

  private var scalikeJdbc: DBsFromConfig = _
  private val createUserTablesFile =
    Paths.get("ddl-scripts/create_user_tables.sql").toFile

  def setupScalikeJdbcConnectionPool(config: Config): Unit = {
    scalikeJdbc = DBsFromConfig.fromConfig(config)
  }

  def closeScalikeJdbcConnectionPool(): Unit = {
    scalikeJdbc.closeAll()
  }

  def dropAndRecreateTables(system: ActorSystem[_]): Unit = {
    implicit val sys: ActorSystem[_] = system
    implicit val ec: ExecutionContext = system.executionContext

    // ok to block here, main thread
    Await.result(
      for {
        _ <- SchemaUtils.dropIfExists()
        _ <- SchemaUtils.createIfNotExists()
        _ <- JdbcProjection.dropOffsetTableIfExists(() =>
          new ScalikeJdbcSession())
        _ <- JdbcProjection.createOffsetTableIfNotExists(() =>
          new ScalikeJdbcSession())
      } yield Done,
      30.seconds)
    if (createUserTablesFile.exists()) {
      Await.result(
        for {
          _ <- dropUserTables()
          _ <- SchemaUtils.applyScript(createUserTablesSql)
        } yield Done,
        30.seconds)
    }
    LoggerFactory
      .getLogger("shopping.cart.CreateTableTestUtils")
      .info("Created tables")
  }

  private def dropUserTables()(
      implicit system: ActorSystem[_]): Future[Done] = {
    SchemaUtils.applyScript("DROP TABLE IF EXISTS public.item_popularity;")
  }

  private def createUserTablesSql: String = {
    val source = scala.io.Source.fromFile(createUserTablesFile)
    val contents = source.mkString
    source.close()
    contents
  }
}
